src/V3Number.h

```diff
@@ -57,53 +56,308 @@ class V3NumberData final {
         }
     };
 
+    enum class V3NumberDataType : uint8_t {
+        UNINITIALIZED = 0,
+        LOGIC = 1,
+        DOUBLE = 2,
+        STRING = 3,
+    };
+    inline friend std::ostream& operator<<(std::ostream& os, const V3NumberDataType& rhs) {
+        switch (rhs) {
+        case V3NumberDataType::UNINITIALIZED: return os << "UNINITIALIZED";
+        case V3NumberDataType::LOGIC: return os << "LOGIC";
+        case V3NumberDataType::DOUBLE: return os << "DOUBLE";
+        case V3NumberDataType::STRING: return os << "STRING";
+        }
+        return os;
+    }
+
 private:
+    // CONSTANTS
+    // At least 2 words (64 fourstate bits). 4 words (128 fourstate bits) in most cases,
+    // i.e. when std::string has 32 bytes.
+    static constexpr int INLINE_WORDS
+        = vlstd::max(2ul, sizeof(std::string) / sizeof(ValueAndX),
+                     sizeof(std::vector<ValueAndX>) / sizeof(ValueAndX));
+    // When m_width > MAX_INLINE_WIDTH number is stored in m_dynamicNumber.
+    // Otherwise number is stored in m_inlineNumber.
+    static constexpr int MAX_INLINE_WIDTH = INLINE_WORDS * sizeof(ValueAndX) / 2 * 8;
+
     // MEMBERS
-    static constexpr size_t m_inlinedSize = 2;  // Can hold 64 bit without dynamic allocation
-    ValueAndX m_inlined[m_inlinedSize];  // Holds the beginning m_inlinedSize words
-    std::vector<ValueAndX> m_data;  // Holds m_inlinedSize-th word and later
+    union {
+        std::array<ValueAndX, INLINE_WORDS> m_inlineNumber;
+        std::vector<ValueAndX> m_dynamicNumber;
+        std::string m_string;
+    };
+
+    int m_width = 0;  // Width (in bits) as specified/calculated
+    V3NumberDataType m_type;
+
+public:
+    bool m_sized : 1;  // True if the user specified the width, else we track it.
+    bool m_signed : 1;  // True if signed value
+    bool m_isNull : 1;  // True if "null" versus normal 0
+    bool m_fromString : 1;  // True if from string literal
+    bool m_autoExtend : 1;  // True if SystemVerilog extend-to-any-width
 
 public:
     // CONSTRUCTORS
-    V3NumberData() {
-        for (size_t i = 0; i < m_inlinedSize; ++i) m_inlined[i] = {0, 0};
+    V3NumberData()
+        : m_type{V3NumberDataType::UNINITIALIZED}
+        , m_sized{false}
+        , m_signed{false}
+        , m_isNull{false}
+        , m_fromString{false}
+        , m_autoExtend{false} {}
+
+    ~V3NumberData() {
+        if (isDynamicNumber())
+            destroyDynamicNumber();
+        else if (isString())
+            destroyString();
```

Comment:
```suggestion
        if (isDynamicNumber()) {
            destroyDynamicNumber();
        } else if (isString()) {
            destroyString();
        }
```

Similar other places, but given I see this three places it's probably best to make a generic "destroyValue()" to remove the repetition.

Reply:
done